pkg/core/policy/dataplane_matcher_test.go

```diff
@@ -0,0 +1,384 @@
+package policy_test
+
+import (
+	"sort"
+	"time"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+
+	mesh_proto "github.com/Kong/kuma/api/mesh/v1alpha1"
+	"github.com/Kong/kuma/pkg/core/policy"
+	mesh_core "github.com/Kong/kuma/pkg/core/resources/apis/mesh"
+	model "github.com/Kong/kuma/pkg/core/xds"
+
+	test_model "github.com/Kong/kuma/pkg/test/resources/model"
+)
+
+var _ = Describe("Dataplane matcher", func() {
+
+	Describe("DataplanePolicyByName", func() {
+
+		type testCase struct {
+			input    []policy.DataplanePolicy
+			expected []policy.DataplanePolicy
+		}
+
+		DescribeTable("should sort DataplanePolicy by Name",
+			func(given testCase) {
+				// when
+				sort.Stable(policy.DataplanePolicyByName(given.input))
+				// then
+				Expect(given.input).To(ConsistOf(given.expected))
+			},
+			Entry("DataplanePolicy in the same mesh", testCase{
+				input: []policy.DataplanePolicy{
+					&mesh_core.ProxyTemplateResource{
+						Meta: &test_model.ResourceMeta{
+							Mesh: "demo",
+							Name: "last",
+						},
+					},
+					&mesh_core.ProxyTemplateResource{
+						Meta: &test_model.ResourceMeta{
+							Mesh: "demo",
+							Name: "first",
+						},
+					},
+				},
+				expected: []policy.DataplanePolicy{
+					&mesh_core.ProxyTemplateResource{
+						Meta: &test_model.ResourceMeta{
+							Mesh: "demo",
+							Name: "first",
+						},
+					},
+					&mesh_core.ProxyTemplateResource{
+						Meta: &test_model.ResourceMeta{
+							Mesh: "demo",
+							Name: "last",
+						},
+					},
+				},
+			}),
+		)
+	})
+
+	Describe("SelectDataplanePolicy()", func() {
+
+		type testCase struct {
+			proxy    *model.Proxy
+			policies []policy.DataplanePolicy
+			expected policy.DataplanePolicy
+		}
+
+		DescribeTable("should find the best match (the one with the highest number of matching key-value pairs)",
+			func(given testCase) {
+				// when
+				actual := policy.SelectDataplanePolicy(given.proxy.Dataplane, given.policies)
+				// then
+				if given.expected == nil {
```

Comment:
Apparently it is not
```
      Refusing to compare <nil> to <nil>.
      Be explicit and use BeNil() instead.  This is to avoid mistakes where both sides of an assertion are erroneously uninitialized.
```